zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,192 @@
+            throw new EndpointGroupException(
+                    "failed to connect to ZooKeeper:  " + zkConnectionStr + " with error:  " + e.getMessage());
+        }
+    }
+
+    /**
+     * Create a Zookeeper endpoint group with a {@link DefaultZkNodeValueConverter}.
+     * @param zkConnectionStr A connection string containing a comma
+     *                          separated list of host:port pairs , each corresponding to a ZooKeeper server
+     * @param zNode a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout   Zookeeper session timeout in milliseconds
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout) {
+        this(zkConnectionStr, zNode, sessionTimeout, new DefaultZkNodeValueConverter());
+    }
+
+    @Override
+    public List<Endpoint> endpoints() {
+        return endpointList;
+    }
+
+    final class ZkWatcher implements Watcher, StatCallback {
+        private int retryCounter = 1;
+
+        @Override
+        public void process(WatchedEvent event) {
+            String path = event.getPath();
+            if (event.getType() != Event.EventType.None) {
+                if (path != null && path.equals(zNode)) {
+                    // Something has changed on the node, let's find out
+                    zooKeeper.exists(zNode, true, this, null);
+                }
+            }
+        }
+
+        @Override
+        public void processResult(int responseCodeInt, String path, Object ctx, Stat stat) {
+            boolean exists;
+            Code responseCode = get(responseCodeInt);
+            switch (responseCode) {
+                case OK:
+                    exists = true;
+                    break;
+                case NONODE:
+                    exists = false;
+                    break;
+                case SESSIONEXPIRED:
+                case NOAUTH:
+                    return;
+                default:
+                    // Retry errors ,by using a exponential backoff strategy
+                    if (retryCounter < MAX_ATTEMPT_COUNT) {
+                        try {
+                            Thread.sleep(retryCounter * 1000);
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                        retryCounter++;
+                        zooKeeper.exists(zNode, true, this, null);
+                    } else { // reset the counter to restart the attempt ,
+                        // we still believe it is a temporary error
+                        retryCounter = 1;
+                    }
+                    return;
+            }
+            byte[] nodeData = null;
+            if (exists) {
+                try {
+                    nodeData = zooKeeper.getData(zNode, false, null);
+                } catch (Exception e) {
+                    logger.warn("error to get zNode data:" + e.getMessage());
```

Comment:
this <getData> method should not be retried, as abnormal exceptions are already cought by the previous <switch> clause, here is just a common get data operation .